### PR TITLE
Add Tip of the Week #147

### PR DESCRIPTION
**Pull request by asoffer**
Use Exhaustive `switch` Statements Responsibly

### CODE REVIEW

File: _posts/2019-09-06-totw-147.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    can take advantage of exhaustive `switch` statements.
+*   Explicitly reserve the right to add new enumerators without notice, to
+    discourage consumers from writing exhaustive `switch` statements. One
+    idiomatic way of doing so is to add a sentinel enumerator clearly not meant
+    to be used in API consumers' exhaustive switch statements; e.g.
+    `kNotForUseWithExhaustiveSwitchStatements`.
+
+## FAQ
+
+*   Why does the compiler allow omitting a `return` statement after an
```
Comment (manshreck) [inline _posts/2019-09-06-totw-147.md]: Markdown is terrible for bulleted lists that have anything but the most basic format. (Additionally, our CSS causes bullets followed by whitespace anywhere in the list to show up alone on their own line. This is a bug bu alas.)

I don't think these should really be list items anyway. The bulleted parts can be replaced with bold text, IMO. (Enclose in ** ... **

File: _posts/2019-09-06-totw-147.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
+---
+title: "Tip of the Week #147: Use Exhaustive `switch` Statements Responsibly"
```
[the remark this answers is not in the document]
Comment (asoffer) [reply _posts/2019-09-06-totw-147.md]: Done.

File: _posts/2019-09-06-totw-147.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+layout: tips
+sidenav: side-nav-tips.html
+published: true
+permalink: tips/147
+type: markdown
+order: "147"
+---
+
+Originally published as totw/147 on 2018-04-25
```
[the remark this answers is not in the document]
Comment (asoffer) [reply _posts/2019-09-06-totw-147.md]: Done.

File: _posts/2019-09-06-totw-147.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+`enum` type without a `default` label will fail to compile if any enumerator of
+the `enum` doesn't have a corresponding `case`. This is sometimes called an
+*exhaustive* or *defaultless* `switch` statement.
+
+An exhaustive `switch` statement is an excellent construct for ensuring at
+compile time that every enumerator of a given enum is explicitly handled.
+However, we must ensure that we handle the fall-through case when the variable
+(legally!) has a non-enumerator value, and one of:
+
+1.  The `enum`-owner guarantees no new enumerators will be added,
```
[the remark this answers is not in the document]
Comment (asoffer) [reply _posts/2019-09-06-totw-147.md]: Done.

File: _posts/2019-09-06-totw-147.md (hunk 1 above)
Comment (asoffer) [reply _posts/2019-09-06-totw-147.md]: Done.

File: _posts/2019-09-06-totw-147.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    to update our `switch` statements when updating to the latest version of the
+    `enum`-owner's code.
+
+### An Initial Attempt
+
+Suppose we are writing a function that maps each enumerator of an `enum` to a
+`std::string`. We decide to use an exhaustive switch statement to ensure we
+didn't forget to handle any of the enumerators:
+
+```c++ {.bad}
```
[the remark this answers is not in the document]
Comment (asoffer) [reply _posts/2019-09-06-totw-147.md]: Done.

**Pull request merged**